llvm/include/llvm/Analysis/TargetTransformInfoImpl.h

```diff
@@ -154,30 +154,22 @@ class TargetTransformInfoImplBase {
   virtual bool isNoopAddrSpaceCast(unsigned, unsigned) const { return false; }
 
   // Assuming that the cast between the two given addrspaces is not a noop,
-  // calculate the width of the given mask value so that it can be applied to
-  // the destination addrspace. In case it cannot be applied since the cast
-  // between the two addrspaces is invalid or the mask value is larger than the
-  // resulting addrspace bit-width, return an empty optional.
-  //
-  // Note that this currently expects the addrspaces to be integral. In case one
-  // of them isn't, an empty optional is returned.
-  virtual std::optional<uint64_t>
-  getAddrSpaceCastMaskWidth(unsigned FromAS, unsigned ToAS, Value *MaskOp,
-                            Instruction *I) const {
+  // calculate the known bits of the resulting ptr in the destination addrspace.
+  // The default implementation returns 0 known bits in case either one of the
+  // addrspaces is not integral or the bit size of the source addrspace is
+  // smaller than the bit size of the destination addrspace.
+  virtual KnownBits computeKnownBitsAddrSpaceCast(unsigned FromAS,
+                                                  unsigned ToAS) const {
     if (DL.isNonIntegralAddressSpace(FromAS) ||
         DL.isNonIntegralAddressSpace(ToAS))
-      return std::nullopt;
-    // All valid 64-bit to 32-bit casts work by chopping off the high
-    // bits. Any masking only clearing the low bits will also apply in the new
-    // address space.
-    if (DL.getPointerSizeInBits(FromAS) != 64 ||
-        DL.getPointerSizeInBits(ToAS) != 32)
-      return std::nullopt;
-    // TODO: Do we need to thread more context in here?
-    KnownBits Known = computeKnownBits(MaskOp, DL, nullptr, I);
-    if (Known.countMinLeadingOnes() < 32)
-      return std::nullopt;
-    return 32;
+      return 0;
+    unsigned FromASBitSize = DL.getPointerSizeInBits(FromAS);
+    unsigned ToASBitSize = DL.getPointerSizeInBits(ToAS);
+    // Per default, we assume that all valid "larger" (e.g. 64-bit) to "smaller"
+    // (e.g. 32-bit) casts work by chopping off the high bits.
+    if (FromASBitSize < ToASBitSize)
+      return 0;
+    return ToASBitSize;
```

Comment:
These implicit conversions from unsigned are broken. I think the KnownBits constructors should have explicit added. 

This should always return a KnownBits with the correct bitwidth and change the mask value, not the bitwidth

Reply:
> I think the KnownBits constructors should have explicit added.

do you dislike the implicit constructor of KnownBits because it's intuitively a bit ambiguous if the unsigned argument stands for "number of known bits" or "bitwidth"?